colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,259 @@
+            loss = closure()
+        """
+        param_groups: Dict
+        {
+            "params":[weight, bias]
+            "lr"
+            "eps"
+            "clip_threshold"
+            "decay_rate"
+            "beta1"
+            "weight_decay"
+            "scale_parameter"
+            "relative_step"
+            "warmup_init"
+        }
+        """
+        for group in self.param_groups:
+            # update weight & bias
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = p.grad
+                if grad.is_sparse:
+                    raise RuntimeError("Adafactor does not support sparse gradients.")
+                state = self.state[p]
+                grad_shape = grad.shape # 1 dim shape
+                param_is_dtensor = is_distributed_tensor(self.shard_to_param.get(id(p)))
+                if param_is_dtensor:
+                    grad_shape = self.shard_to_param.get(id(p)).shape # tp shape (2 dim)
+                else:
+                    grad_shape = grad.shape  # tp and zero shape (1 dim)
+                factored, use_first_moment = self._get_options(group, grad_shape)
+                if len(state) == 0:
+                    state["step"] = 0
+                    if use_first_moment:
+                        # Exponential moving average of gradient values
+                        state["exp_avg"] = torch.zeros_like(grad)
+                    if factored:
+                        shard_spec = get_sharding_spec(self.shard_to_param.get(id(p)))
+                        if shard_spec.sharding_sequence[0] == 'R': # Col Parallel
+                            state["exp_avg_sq_row"] = torch.zeros(grad_shape[0] // self.data_parallel_size).to(grad)  # [H/dp]
+                            state["exp_avg_sq_col"] = torch.zeros(grad_shape[1]).to(grad)  # [W/TP]
+                            
+                        if shard_spec.sharding_sequence[-1] == 'R': # Row Parallel
+                            state["exp_avg_sq_row"] = torch.zeros(grad_shape[0] // self.tensor_parallel_size).to(grad)  # [H/dp/Tp]
+                            state["exp_avg_sq_col"] = torch.zeros(grad_shape[1]).to(grad)  # [W/TP]
+                    else:
+                        state["exp_avg_sq"] = torch.zeros_like(grad)
+                    state["RMS"] = 0
+                else:
+                    if use_first_moment:
+                        state["exp_avg"] = state["exp_avg"].to(grad)
+                    if factored:
+                        state["exp_avg_sq_row"] = state["exp_avg_sq_row"].to(grad)
+                        state["exp_avg_sq_col"] = state["exp_avg_sq_col"].to(grad)
+                    else:
+                        state["exp_avg_sq"] = state["exp_avg_sq"].to(grad)
+                
+                p_data_fp32 = p
+                if p.dtype in {torch.float16, torch.bfloat16}:
+                    p_data_fp32 = p_data_fp32.float()
```

Comment:
AMP logic should not be added into this class

Reply:
> AMP logic should not be added into this class


Should I remove the “if” judgment about AMP or remove all p_data_fp32 related operations?